cmd/outline-ss-server/metrics.go

```diff
@@ -49,9 +62,106 @@ type outlineMetrics struct {
 var _ service.TCPMetrics = (*outlineMetrics)(nil)
 var _ service.UDPMetrics = (*outlineMetrics)(nil)
 
+type ReportTunnelTimeFunc func(IPKey, ipinfo.IPInfo, time.Duration)
+
+type activeClient struct {
+	mu              sync.Mutex
+	IPKey           IPKey
+	clientInfo      ipinfo.IPInfo
+	connectionCount int
+	startTime       time.Time
+}
+
+type IPKey struct {
+	ip        netip.Addr
+	accessKey string
+}
+
+type tunnelTimeTracker struct {
+	mu               sync.Mutex
+	activeClients    map[IPKey]*activeClient
+	reportTunnelTime ReportTunnelTimeFunc
+}
+
+// Reports time connected for all active clients, called at a regular interval.
+func (t *tunnelTimeTracker) reportAll(now time.Time) {
+	if len(t.activeClients) == 0 {
+		logger.Debugf("No active clients. No TunnelTime activity to report.")
+		return
+	}
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	for _, c := range t.activeClients {
```

Comment:
This is a big point of contention. reportAll will block every new connection. I wonder if this may become an issue with several thousand clients.
We had to be careful about that with the [cipher list](https://github.com/Jigsaw-Code/outline-ss-server/blob/master/service/cipher_list.go).

Reply:
What if I put the lock on the individual items in the map (the `activeClient` objects) instead of the entire map. That way it would only blocks connections for individual clients that are trying to access the same `activeClient`, which would be much less contentious. Wdyt?